packages/webdoc-template-library/src/pipeline-elements/Sitemap.js

```diff
@@ -0,0 +1,67 @@
+// @flow
+
+import * as fse from "fs-extra";
+import type {TemplatePipeline, TemplatePipelineElement} from "../TemplatePipeline";
+import path from "path";
+
+type SitemapData = {
+  outputFile: string;
+};
+
+export class Sitemap implements TemplatePipelineElement<SitemapData> {
+  urls: string[] = [];
+  dir: string;
+  domain: string;
+  root: string;
+
+  /**
+   * @param {string} dir - Directory to save sitemap.xml
+   * @param {string} domain - The site domain with protocol
+   * @param {string} root - the siteroot
+   */
+  constructor(dir: string, domain: string, root: string) {
+    this.dir = dir;
+    this.domain = domain;
+    this.root = root;
+
+    if (this.domain.charAt(this.domain.length - 1) !== "/") {
+      this.domain += "/";
+    }
+  }
+
+  attachTo(pipeline: TemplatePipeline) {
+    // noop
+  }
+
+  run(input: string, pipelineData: SitemapData = {}) {
+    if (pipelineData.outputFile) {
+      this.urls.push(pipelineData.outputFile);
+    }
+  }
+
+  close() {
+    /* eslint-disable max-len */
+    const xml =
+`<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
+   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+   xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
+${this.urls.map((url) =>
+    `  <url>
+    <loc>${this.domain + path.join(this.root, path.relative(this.dir, url))}</loc>
+  </url>`).join("\n")}
+</urlset>`;
+    /* eslint-enable max-len */
```

Comment:
Wouldn't it be better to make another _template_ file instead of hardcoding generation?

Reply:
I'm not sure what you mean by hardcoding generation. The template package can still exclude sitemap generation by simply not adding this pipeline element.

Using a pipeline element instead of a template means you don't have to worry about collecting the URLs that were generated. (Documents are the only generated URL - tutorial, class index, package READMEs are all generated)

Reply:
I meant this exact part of code could be stored in separated file instead of defining it right in the code. (So you could change XML template with ease.)